src/bandwidth.rs

```diff
@@ -0,0 +1,271 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::{Multiaddr, core::Transport, core::transport::TransportError};
+use futures::{prelude::*, try_ready};
+use lazy_static::lazy_static;
+use parking_lot::Mutex;
+use smallvec::{smallvec, SmallVec};
+use std::{cmp, io, io::Read, io::Write, sync::Arc, time::Duration, time::Instant};
+
+/// Wraps around a `Transport` and logs the bandwidth that goes through all the opened connections.
+#[derive(Clone)]
+pub struct BandwidthLogging<TInner> {
+    inner: TInner,
+    sinks: Arc<BandwidthSinks>,
+}
+
+impl<TInner> BandwidthLogging<TInner> {
+    /// Creates a new `BandwidthLogging` around the transport.
+    #[inline]
+    pub fn new(inner: TInner, period: Duration) -> (Self, Arc<BandwidthSinks>) {
+        let mut period_seconds = cmp::min(period.as_secs(), u64::from(u32::max_value())) as u32;
```

Comment:
How about just capping the period to one POSIX day?

```suggestion
        let mut period_seconds = cmp::min(period.as_secs(), 86400) as u32;
```